source/common/websocket/codec.cc

```diff
@@ -0,0 +1,215 @@
+  frameMaskFlag(flag_and_length);
+  if (length_ == 0x7e) {
+    length_of_extended_length_ = kPayloadLength16Bit;
+    length_ = 0;
+    state_ = State::FrameHeaderExtendedLength;
+  } else if (length_ == 0x7f) {
+    length_of_extended_length_ = kPayloadLength64Bit;
+    length_ = 0;
+    state_ = State::FrameHeaderExtendedLength;
+  } else if (masking_key_length_ > 0) {
+    state_ = State::FrameHeaderMaskingKey;
+  } else {
+    frameDataStart(input, output);
+  }
+}
+
+void Decoder::doDecodeExtendedLength(const uint8_t length, Buffer::Instance& input,
+                                     absl::optional<std::vector<Frame>>& output) {
+  if (length_of_extended_length_ == 1) {
+    length_ |= static_cast<uint64_t>(length);
+  } else {
+    length_ |= static_cast<uint64_t>(length) << 8 * (length_of_extended_length_ - 1);
+  }
+  length_of_extended_length_--;
+  if (length_of_extended_length_ == 0) {
+    if (masking_key_length_ > 0) {
+      state_ = State::FrameHeaderMaskingKey;
+    } else {
+      frameDataStart(input, output);
+    }
+  }
+}
+
+void Decoder::doDecodeMaskingKey(const uint8_t mask, Buffer::Instance& input,
+                                 absl::optional<std::vector<Frame>>& output) {
+  if (!frame_.masking_key_.has_value()) {
+    frame_.masking_key_ = 0;
+  }
+  if (masking_key_length_ == 1) {
+    frame_.masking_key_.value() |= static_cast<uint32_t>(mask);
+  } else {
+    frame_.masking_key_.value() |= static_cast<uint32_t>(mask) << 8 * (masking_key_length_ - 1);
+  }
+  masking_key_length_--;
+  if (masking_key_length_ == 0) {
+    frameDataStart(input, output);
+  }
+}
+
+void Decoder::doDecodePayload(const uint64_t slice_length, const uint8_t*& mem,
+                              uint64_t& slice_index, Buffer::Instance& input,
+                              absl::optional<std::vector<Frame>>& output) {
+  uint64_t remain_in_buffer = slice_length - slice_index;
+  if (remain_in_buffer <= length_) {
+    frameData(mem, remain_in_buffer);
+    mem += remain_in_buffer;
+    slice_index += remain_in_buffer;
+    bytes_consumed_by_frame_ += remain_in_buffer;
+    length_ -= remain_in_buffer;
+  } else {
+    frameData(mem, length_);
+    mem += length_;
+    slice_index += length_;
+    bytes_consumed_by_frame_ += length_;
+    length_ = 0;
+  }
+  if (length_ == 0) {
+    frameDataEnd(input, output);
+    state_ = State::FrameHeaderFlagsAndOpcode;
+  }
+}
+
+absl::optional<std::vector<Frame>> Decoder::decode(Buffer::Instance& input) {
+  absl::optional<std::vector<Frame>> output = absl::nullopt;
+  if (!inspect(input, output)) {
+    return absl::nullopt;
+  }
+  return output;
+}
+
+bool Decoder::inspect(Buffer::Instance& input, absl::optional<std::vector<Frame>>& output) {
+  for (const Buffer::RawSlice& slice : input.getRawSlices()) {
+    const uint8_t* mem = reinterpret_cast<uint8_t*>(slice.mem_);
+    for (uint64_t slice_index = 0; slice_index < slice.len_;) {
+      uint8_t byte_of_frame = *mem;
```

Comment:
There are quite a few local variables here which need to be tracked. How about something like:
```
  for (const Buffer::RawSlice& slice : input.getRawSlices()) {
    absl::Span<const uint8_t> data(reinterpret_cast<uint8_t*>(slice.mem_), slice_len_);
    while (!data.empty()) {
      uint64_t bytes_decoded = 0;
      switch (state_) {
      case State::FrameHeaderFlagsAndOpcode:
        bytes_decoded = doDecodeFlagsAndOpcode(data);
        if (bytes_decoded = 0) {
          return false;
        }
        break;
      case State::FrameHeaderMaskFlagAndLength:
        bytes_decoded = doDecodeMaskFlagAndLength(data);
        break;
      case State::FrameHeaderExtendedLength:
        bytes_decoded = doDecodeExtendedLength(data);
        break;
      case State::FrameHeaderMaskingKey:
        bytes_decoded = doDecodeMaskingKey(data);
        break;
      case State::FramePayload:
        bytes_decoded = doDecodePayload(data);
        break;
      case State::FrameFinished:
        doFrameEnd(input, output);
        break;
      }
      data.remove_prefix(bytes_decoded);
      bytes_consumed_by_frame_ += bytes_decoded;
```

Notice that all of the `doDecodeFoo()` methods now simply take the span, and return the number of bytes they processed. The exception is `doFrameEnd()` which drains the input, adds the frame to the output, and resets state.